File: Cc.Core/Models/Principal.cs
using System;
using System.Collections.Generic;

namespace Cc.Core.Models;

public partial class Principal
{
    public long Id { get; set; }

    public string? InsertUser { get; set; }

    public DateTime InsertDate { get; set; }

    public string? UpdateUser { get; set; }

    public DateTime UpdateDate { get; set; }

    public string Username { get; set; } = null!;

    public string? Password { get; set; }

    public string? Name { get; set; }

    public string? Surname { get; set; }

    public string? Mail { get; set; }

    public string? Phone { get; set; }

    public string? Role { get; set; }

    public bool Disabled { get; set; }

    public string? TaxCode { get; set; }

    public long? CompanyId { get; set; }

    public string? Language { get; set; }

    public long? AvatarId { get; set; }

    public string? City { get; set; }

    public string? Mobile { get; set; }

    public string? District { get; set; }

    public string? Cap { get; set; }

    public string? Notes { get; set; }

    public bool AgendaLocked { get; set; }

    public string? Country { get; set; }

    public string? Address { get; set; }

    public virtual ICollection<Appointment> Appointment { get; set; } = new List<Appointment>();

    public virtual Document? Avatar { get; set; }

    public virtual Company? Company { get; set; }

    public virtual ICollection<Customer> Customer { get; set; } = new List<Customer>();

    public virtual ICollection<CustomerCall> CustomerCall { get; set; } = new List<CustomerCall>();

    public virtual ICollection<PrincipalAuth> PrincipalAuth { get; set; } = new List<PrincipalAuth>();

    public virtual ICollection<PrincipalFather> PrincipalFatherFather { get; set; } = new List<PrincipalFather>();

    public virtual ICollection<PrincipalFather> PrincipalFatherPrincipal { get; set; } = new List<PrincipalFather>();

    public virtual ICollection<PrincipalGrouping> PrincipalGrouping { get; set; } = new List<PrincipalGrouping>();

    public virtual ICollection<PrincipalRole> PrincipalRole { get; set; } = new List<PrincipalRole>();

    public virtual ICollection<PrincipalTabRegion> PrincipalTabRegion { get; set; } = new List<PrincipalTabRegion>();

    public virtual ICollection<Warehouse> Warehouse { get; set; } = new List<Warehouse>();
}

File: Cc.Ui/src/models/principal.ts
import { Entity, resource, field, required, association, regex, regexErrorMessage } from '../framework/entity';
import { Pagination, Pager } from '../framework/core/actions';
import { PrincipalAuth } from './principalAuth';
import { Company } from './company';
import { cache } from '../app/cache';
import { LookupKey } from '../core/lookup';
import {PrincipalTabRegion} from './principalTabRegion';
import { Appointment } from './appointment';

@resource('principal')
export class Principal extends Entity {
    
    @field()
    id: number;
    @field()
    insertDate: string;
    @field()
    insertUser: string;
    @field()
    updateDate: string;
    @field()
    updateUser: string;
    @field()
    companyId: number;
    @field() @required()
    surname: string;
    @field() @required()
    name: string;
    @field() @required()
    taxCode: string;
    @field() @required()
    mail: string;
    @field()
    phone: string;
    @field()
    mobilePhone: string;
    @field() @required()
    username: string;
    @field() 
    password: string;
    @field() @required()
    role: string;
    @field()
    language?: string;
    @field()
    avatarId?: number;
    @field()
    disabled: boolean;
    @field()
    agendaLocked: boolean;
    @field()
    country: string;
    @field()
    province: string;
    @field()
    city: string;
    @field()
    address: string;
    @field()
    cap: string;
    @field()
    notes: string;
    
    error: string;
    token: string;
    agencyIds: number[];
    productIds: number[];

    @association(Company, 'companyId', (id: string | number) => cache(LookupKey.COMPANY, id))
    company: Company;
    @association(PrincipalAuth)
    principalAuth: PrincipalAuth[];
    @association(PrincipalTabRegion)
    principalTabRegion: PrincipalTabRegion[];
    @association(Appointment)
    appointment: Appointment[];
    
    constructor(data?: any) {
        super()
        this.init(data);      
    }

    static newItem(): Principal {

        let item = new Principal();
        return item;
    }
}

@resource('principal')
export class PrincipalSearchModel extends Entity implements Pagination {
    
    @field()
    companyId?: number;
    @field()
    surname?: string;
    @field()
    name?: string;
    @field()
    taxCode?: string;
    @field()
    mail?: string;
    @field()
    phone?: string;
    @field()
    username?: string;
    @field()
    password?: string;
    @field()
    role?: string;
    @field()
    disabled?: boolean;

    @field()
    pager: Pager = <Pager> {
        take: 20,
        orderBy: 'surname'
    };
}

export class PrincipalUpload extends Entity {

    @field()
    companyId: number;
    @field()
    fileEncoded: string;
    @field()
    filename: string;
    @field()
    notify: boolean;
}

@resource('principalUploadResult')
export class PrincipalUploadResult extends Entity {

    @field()
    error: string;
    @field()
    corrects: number;
    @field()
    usersWithErrors: Principal[] = [];
    @field()
    createdResource: string[] = [];

    get wrongs() {
        return (this.usersWithErrors || []).length;
    }
    
    constructor(data?: any) {
        super()
        this.init(data);      
    }

}
